staging/src/k8s.io/apiserver/pkg/storage/cacher/btree_store.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/google/btree"
+	"k8s.io/client-go/tools/cache"
+)
+
+type btreeStore struct {
+	lock sync.RWMutex
+	tree *btree.BTree
+	n    int
+}
+
+func newBtreeStore(degree int) *btreeStore {
+	return &btreeStore{
+		tree: btree.New(degree),
+	}
+}
+
+func (t *btreeStore) Add(obj interface{}) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	return t.addOrUpdateLocked(obj, true)
+}
+
+func (t *btreeStore) Update(obj interface{}) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	return t.addOrUpdateLocked(obj, false)
+}
+
+func (t *btreeStore) Delete(obj interface{}) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	storeElem, ok := obj.(storeElement)
+	if !ok {
+		return fmt.Errorf("obj not a storeElement: %#v", obj)
+	}
+	item := t.tree.Delete(storeElem)
+	if item == nil {
+		return fmt.Errorf("obj does not exist")
+	}
+	t.n--
+
+	return nil
+}
+
+func (t *btreeStore) List() []interface{} {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	items := make([]interface{}, 0, t.n)
```

Comment:
I don't think you really need to maintain `t.n` here - you can just use btree.Len(), right?

Reply:
right